va/va_prot.h

```diff
@@ -0,0 +1,379 @@
+ * vaCreateProtectedSession().
+ *
+ * The general control flow is demonstrated by the following pseudo-code:
+ * \code
+ * // Create config
+ * VAConfigID config_id;
+ *
+ * wired.bits.session_type = VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0;
+ * wired.bits.protocol_version = VA_PROTECTED_DISPLAY_LINK_HDCP_VERSION_22;
+ * attrib_prot[0].value = wired.value;
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected, VAEntrypointProtectedDisplayLink, attrib_prot, 1, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected session via vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID protected_session;
+ *
+ * va_status = vaCreateProtectedSession(
+ *     va_dpy,
+ *     config_id,
+ *     &protected_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_prot_attach Attach/Detach protected session to the context which
+ * want to enable/disable protection
+ *
+ * Protected session is attached to decode/encode/vp context to enable protection.
+ *
+ * \code
+ * vaAttachProtectedSession(va_dpy, decode_ctx, protected_session);
+ * foreach (iteration) {
+ *     vaBeginPicture(va_dpy, decode_ctx, surface);
+ *     ...
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
+ *     ...
+ *     vaEndPicture(va_dpy, decode_ctx);
+ * }
+ * vaDetachProtectedSession(va_dpy, decode_ctx);
+ * \endcode
+ *
+ * \section api_prot_update Updating protected session status related to
+ * hardware settings
+ *
+ * A function for updating hardware status of the protected session. Sometimes
+ * the status is initialized at protected session creation time. When the
+ * information is gathered, we can update the protected session status based
+ * on the information we got with vaProtectedSessionHwUpdate().
+ * For example, if the protected session is for protected content, we will get
+ * policy blobs from TEE and then pass these infomation with
+ * vaProtectedSessionHwUpdate to change the session status to enable specific
+ * hardware session type.
+ *
+ * \code
+ * INFO *info = &oem_policy_blobs;
+ * VABufferID buffer;
+ * VAProtectedSessionHwUpdateBuffer update_buff = {0};
+ *
+ * update_buff.input.data = info;
+ * update_buff.input.data_size = sizeof(info);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionHwUpdateBufferType,
+ *                 sizeof(update_buff),
+ *                 1,
+ *                 &update_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionHwUpdate(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ *
+ * \section api_prot_exec TEE communication via vaProtectedSessionExecute()
```

Comment:
Can we, please, change a document style? You describe some API sections like `api_prot_exec`, but effectively you fall into providing function reference. But you actually have function reference when you declare each function!

I suggest to completely remove API sections and instead tell a complete story. It seems that vaProtectedSessionHwUpdate and vaProtectedSessionExecute are supposed to be used somewhere between protected session initialization and actual decoding - describe them there.

Reply:
Thanks for your feedback. I restructure and refine the comments and updated. Please check if it is better or not.